File: scala/crypto-square/src/main/scala/CryptoSquare.scala
import scala.annotation.tailrec

object CryptoSquare {

  // needed for the tests to compile and run
  def apply() = this

  type PlainText = String
  type NormalizedPlainText = String
  type SquareSize = Int
  type TextSegments = Seq[String]
  type NormalizedCipherText = String
  type CipherText = String

  def normalizePlaintext(text: PlainText): NormalizedPlainText =
    text filter (_.isLetterOrDigit) map (_.toLower)

  def squareSize(text: PlainText): SquareSize = {
    import math._
    val textLength = normalizePlaintext(text).length
    ceil(sqrt(textLength)) toInt
  }

  private val EmptyPlainTextSegments = List("")

  def plaintextSegments(text: PlainText): TextSegments = {
    val normalizedPlainText = normalizePlaintext(text)

    if (normalizedPlainText.isEmpty) EmptyPlainTextSegments
    else normalizedPlainText grouped squareSize(normalizedPlainText) toList
  }

  private def cipherTextSegments(text: PlainText): TextSegments = {
    def addRowChar(col: Int)(cipherText: CipherText, plainTextRow: PlainText): CipherText =
      plainTextRow.lift(col) map (cipherText + _) getOrElse cipherText

    val segments = plaintextSegments(text)
    val squareColumns = (0 until squareSize(text))
    squareColumns map { col =>
      segments.foldLeft("")(addRowChar(col))
    }
  }

  def normalizedCiphertext(text: PlainText): NormalizedCipherText =
    cipherTextSegments(text) mkString(" ")

  def ciphertext(text: PlainText): CipherText =
    cipherTextSegments(text) mkString
}

File: haskell/crypto-square/CryptoSquare.hs
module CryptoSquare (normalizePlaintext, squareSize, plaintextSegments,
                     ciphertext, normalizeCiphertext) where

import Data.Char
import Data.List.Split (chunksOf)

type PlainText = String
type NormalizedPlainText = String
type SquareSize = Int
type TextSegments = [String]
type NormalizedCipherText = String
type CipherText = String

normalizePlaintext :: PlainText -> NormalizedPlainText
normalizePlaintext =
  map toLower . filter isAlphaNum

squareSize :: PlainText -> SquareSize
squareSize =
 ceiling . sqrt . (fromIntegral :: Int -> Double) . length . normalizePlaintext

plaintextSegments :: PlainText -> TextSegments
plaintextSegments plainText =
  chunksOf (squareSize plainText) normalizedPlainText
  where normalizedPlainText = normalizePlaintext plainText

ciphertext :: PlainText -> CipherText
ciphertext = concat . cipherTextSegments

normalizeCiphertext :: PlainText -> NormalizedCipherText
normalizeCiphertext = unwords . cipherTextSegments

cipherTextSegments :: PlainText -> TextSegments
cipherTextSegments plainText =
  map foldSegments squareColumns
  where
    textSegments = plaintextSegments plainText
    squareColumns = [0..pred $ squareSize plainText]

    foldSegments :: Int -> CipherText
    foldSegments col = foldl (addRowChar col) "" textSegments

    addRowChar :: Int -> CipherText -> PlainText -> CipherText
    addRowChar col cipherText plainTextRow
      | col < length plainTextRow = cipherText ++ [plainTextRow !! col]
      | otherwise = cipherText
